### ISSUE TITLE
Add ParameterNumber.ignoreAnnotations property

### ISSUE THREAD

**Issue by 2is10** [checkstyle/checkstyle#6772]
My company uses [Jackson](https://github.com/FasterXML/jackson) for JSON serialization/deserialization of value classes.

We normally use `@JsonCreator` on constructors to support automatic deserialization.

```java
@Immutable
class Thing {
    private final int id;
    private final String name;
 
    @JsonCreator
    public Thing(@JsonProperty("id") int id, 
                 @JsonProperty("name") String name) {
        this.id = id;
        this.name = name;
    }
}
```

Sometimes beans get large (lots of properties), so we have ParameterNumber configured to ignore constructors using:

```xml
<property name="tokens" value="METHOD_DEF"/>
```

However, sometimes we need to do more validation when deserializing JSON than we do when constructing directly. In these cases, we put @JsonCreator on a factory method that calls the simpler constructor.

```java
    public Thing(int id, String name) {
        this.id = id;
        this.name = Preconditions.checkNotNull(name, "'name' is required");
    }

    @JsonCreator
    private static Thing fromJson(@JsonProperty("id") int id, 
                                  @JsonProperty("name") String name) {
        Preconditions.checkArgument(id > 0, "'id' must be positive");
        Preconditions.checkArgument(!name.isEmpty(), "'name' cannot be empty");
        Preconditions.checkArgument(isValidName(name),
                                    "'name' must consist of only letters and digits");
        // ...
        return new Thing(id, name);
    }
```

Note that the @JsonCreator factory method has the same argument list as the constructor, so we can’t just define another constructor (unless we play with argument order, which could be confusing).

If the number of properties is too large, we have a checkstyle violation. We’d like to be able to configure checkstyle to ignore methods (and constructors) annotated with `@JsonCreator`, much like how `ignoreOverrides` ignores methods annotated with `@Override` (#66).

I’d suggest generalizing by accepting an annotation list (a String Set like `IllegalImport`’s `illegalClasses` property). Maybe it could be called `ignoreAnnotations` or `excludeAnnotations`.

Thanks for considering it.

**Comment by romani** [comment]
Can you try to use https://checkstyle.org/config_filters.html#SuppressionXpathFilter , this filter supports ParameterNumber , so you can suppress violation with any annotation you like.
Search for `Next example is about suppressing method with certain annotation by its name and element value.`

If you are limited to single file configuration you can try https://checkstyle.org/config_filters.html#SuppressionXpathSingleFilter

**Comment by 2is10** [comment]
Wow! Thanks for pointing out `SuppressionXpathFilter`. It looks promising.

I’ve just tried using it. (We’re on checkstyle 8.16.) I added a <suppress-xpath> element to our checkstyle-suppressions.xml file, updated its doctype, and it validates. I also changed our "SuppressionFilter" module to a "SuppressionXpathFilter" module in our checkstyle.xml, but I get:

```
com.puppycrawl.tools.checkstyle.api.CheckstyleException: SuppressionXpathFilter is not allowed as a child in Checker
        at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:483)
        at com.puppycrawl.tools.checkstyle.api.AutomaticBean.configure(AutomaticBean.java:198)
        at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.createRootModule(CheckstyleAntTask.java:412)
```

I’ll go ahead and close this issue since it looks like you’ve already got the functionality I requested, even if I haven’t gotten it working yet. Thanks again.

**Comment by rnveach** [comment]
@2is10 
> We’re on checkstyle 8.16

Filter is only available on 8.18+. See https://checkstyle.sourceforge.io/config_filters.html#SuppressionXpathSingleFilter

**Comment by 2is10** [comment]
I’m using [SuppressionXpathFilter](https://checkstyle.sourceforge.io/config_filters.html#SuppressionXpathFilter), which is documented as:

> Since Checkstyle 8.6

I can try upgrading a little later.
